images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go

```diff
@@ -0,0 +1,154 @@
+package filesystem
+
+import (
+	"crypto/tls"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"k8s.io/klog/v2"
+
+	"github.com/deckhouse/virtualization-controller/pkg/tls/util"
+)
+
+type FileCertificateManager struct {
+	stopCh             chan struct{}
+	certAccessLock     sync.Mutex
+	stopped            bool
+	cert               *tls.Certificate
+	certBytesPath      string
+	keyBytesPath       string
+	errorRetryInterval time.Duration
+}
+
+func NewFileCertificateManager(certBytesPath, keyBytesPath string) *FileCertificateManager {
+	return &FileCertificateManager{
+		certBytesPath:      certBytesPath,
+		keyBytesPath:       keyBytesPath,
+		stopCh:             make(chan struct{}, 1),
+		errorRetryInterval: 1 * time.Minute,
+	}
+}
+
+func (f *FileCertificateManager) Start() {
+	objectUpdated := make(chan struct{}, 1)
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		klog.Errorf("failed to create an inotify watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	certDir := filepath.Dir(f.certBytesPath)
+	err = watcher.Add(certDir)
+	if err != nil {
+		klog.Errorf("failed to establish a watch on %s: %v", f.certBytesPath, err)
+	}
+	keyDir := filepath.Dir(f.keyBytesPath)
+	if keyDir != certDir {
+		err = watcher.Add(keyDir)
+		if err != nil {
+			klog.Errorf("failed to establish a watch on %s: %v", f.keyBytesPath, err)
+		}
+	}
+
+	go func() {
+		for {
+			select {
+			case _, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				select {
+				case objectUpdated <- struct{}{}:
+				default:
+					klog.V(5).Infof("Dropping redundant wakeup for cert reload")
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				klog.ErrorS(err, fmt.Sprintf("An error occurred when watching certificates files %s and %s", f.certBytesPath, f.keyBytesPath))
+			}
+		}
+	}()
+
+	// ensure we load the certificates on startup
+	objectUpdated <- struct{}{}
+
+sync:
+	for {
+		select {
+		case <-objectUpdated:
+			if err := f.rotateCerts(); err != nil {
+				go func() {
+					time.Sleep(f.errorRetryInterval)
+					select {
+					case objectUpdated <- struct{}{}:
+					default:
+						klog.V(5).Infof("Dropping redundant wakeup for cert reload")
+					}
+				}()
+			}
+		case <-f.stopCh:
+			break sync
+		}
+	}
+}
+
+func (f *FileCertificateManager) Stop() {
+	f.certAccessLock.Lock()
+	defer f.certAccessLock.Unlock()
+	if f.stopped {
```

Comment:
No need to use f.stopped

	select {
	case <-f.stopCh:
	default:
		close(f.stopCh)
	}

Reply:
No. if you call Stop more than 1 time you will panic: close of closed channel

Reply:
> No. if you call Stop more than 1 time you will panic: close of closed channel

There will be no panic. case <-f.stopCh will be called (with a closing flag). The default case with the channel closure will not be called.